src/cs/Ssh/SecureStream.cs

```diff
@@ -0,0 +1,291 @@
+	/// opened channel. Additional message processing is kicked off as a background task chain.
+	/// </summary>
+	/// <exception cref="SshConnectionException">The connection failed due to a protocol
+	/// error.</exception>
+	/// <exception cref="TimeoutException">The ConnectTimeout property is set and the initial
+	/// version exchange could not be completed within the timeout.</exception>
+	public async Task ConnectAsync(CancellationToken cancellation = default)
+	{
+		if (Authenticating == null)
+		{
+			throw new InvalidOperationException(
+				"An Authenticating event handler must be registered before connecting.");
+		}
+
+		if (this.serverCredentials != null)
+		{
+			var serverSession = (SshServerSession)this.session;
+			serverSession.Credentials = this.serverCredentials;
+		}
+
+		await this.session.ConnectAsync(this.transportStream, cancellation).ConfigureAwait(false);
+
+		SshChannel? channel = null;
+		if (this.clientCredentials != null)
+		{
+			SshConnectionException? ex = null;
+			var clientSession = (SshClientSession)this.session;
+			if (!(await clientSession.AuthenticateServerAsync(cancellation).ConfigureAwait(false)))
+			{
+				ex = new SshConnectionException(
+					"Server authentication failed.", SshDisconnectReason.HostKeyNotVerifiable);
+			}
+
+			if (ex == null && !(await clientSession.AuthenticateClientAsync(
+				this.clientCredentials, cancellation).ConfigureAwait(false)))
+			{
+				ex = new SshConnectionException(
+					"Client authentication failed.", SshDisconnectReason.NoMoreAuthMethodsAvailable);
+			}
+
+			if (ex != null)
+			{
+				await this.session.CloseAsync(ex.DisconnectReason, ex).ConfigureAwait(false);
+				throw ex;
+			}
+
+			channel = await this.session.OpenChannelAsync(cancellation).ConfigureAwait(false);
+		}
+		else
+		{
+			channel = await this.session.AcceptChannelAsync(cancellation).ConfigureAwait(false);
+		}
+
+		this.stream = CreateStream(channel);
+		channel.Closed += (_, _) =>
+		{
+			this.Dispose();
+		};
+	}
+
+	/// <summary>
+	/// Creates a stream instance for a channel. May be overridden to create a
+	/// <see cref="SshStream" /> subclass.
+	/// </summary>
+	protected virtual SshStream CreateStream(SshChannel channel)
+	{
+		return new SshStream(channel);
+	}
+
+	protected override void Dispose(bool disposing)
+	{
+		if (disposing)
+		{
+			this.stream?.Dispose();
+			this.session.Dispose();
+
+			// If the session has not connected yet, it doesn't know about the stream and won't dispose it.
+			// So we dispose it explicitly here.
+			this.transportStream.Dispose();
+		}
+
+		base.Dispose(disposing);
+	}
+
+	/// <summary>
+	/// Close the SSH session with <see cref="SshDisconnectReason.None"/> reason and dispose the underlying transport stream.
+	/// </summary>
+	public async Task CloseAsync()
+	{
+		await this.session.CloseAsync(SshDisconnectReason.None, this.session.GetType().Name + " disposed").ConfigureAwait(false);
+		this.session.Dispose();
```

Comment:
Should this call `Dispose()`?

Reply:
Disposing the session closes the channel, which will trigger the channel-closed handler that disposes this instance on line 173. I'll add a comment.